### PR TITLE
[libc++] Use proper functions instead of macros in bsd_locale_defaults.h

### PR DESCRIPTION
**Pull request by ldionne**
We were using macros instead of functions, leading to the inability to properly qualify calls to those symbols inside <locale>. This is also a step towards making the locale API modules-correct.

### CODE REVIEW

File: libcxx/include/__locale_dir/locale_base_api/bsd_locale_defaults.h [hunk 1]
```diff
@@ -14,23 +14,102 @@
...
+__libcpp_mbsrtowcs_l(wchar_t* __dest, const char** __src, size_t __len, mbstate_t* __ps, locale_t __loc) {
+  return ::mbsrtowcs_l(__dest, __src, __len, __ps, __loc);
+}
+#endif
+
+template <class... _Args>
+_LIBCPP_HIDE_FROM_ABI int
+__libcpp_snprintf_l(char* __s, size_t __n, locale_t __loc, const char* __format, _Args&&... __args) {
+  _LIBCPP_DIAGNOSTIC_PUSH
+  _LIBCPP_CLANG_DIAGNOSTIC_IGNORED("-Wformat-nonliteral")
```
Comment (philnik777) [inline libcxx/include/__locale_dir/locale_base_api/bsd_locale_defaults.h:89]: Clang supports `__attribute__((format))` on variadic templates.  Since misusing these functions can result in significant security problems, I think it'd be good to add the attributeg.

File: libcxx/include/__locale_dir/locale_base_api.h [hunk 2]
```diff
@@ -46,7 +54,7 @@ size_t __libcpp_mbsnrtowcs_l(wchar_t* dest, const char** src, size_t max_out, si
 size_t __libcpp_mbrtowc_l(wchar_t* dest, cosnt char* src, size_t count, mbstate_t*, locale_t);
 int __libcpp_mbtowc_l(wchar_t* dest, const char* src, size_t count, locale_t);
 size_t __libcpp_mbrlen_l(const char* str, size_t count, mbstate_t*, locale_t);
-lconv* __libcpp_localeconv_l(locale_t);
+lconv* __libcpp_localeconv_l(locale_t&);
```
Comment (philnik777) [inline libcxx/include/__locale_dir/locale_base_api.h:57]: Why is this change required?

File: libcxx/include/__locale_dir/locale_base_api/bsd_locale_defaults.h (hunk 1 above)
Comment (ldionne) [reply libcxx/include/__locale_dir/locale_base_api/bsd_locale_defaults.h:89]: That's right, but GCC doesn't, which triggers `-Wgcc-compat`. Is your preferred approach to add the attribute (only on Clang) and to disable the compatibility warning?

FWIW, I went for the current approach because those are implementation-detail functions not called by users, so I thought the value of the warning was somewhat less. But I'm fine either way, just let me know what you think is best.

Comment (ldionne) [reply libcxx/include/__locale_dir/locale_base_api/bsd_locale_defaults.h:89]: FWIW I think this is the simplest way to write the alternative (this is clang-formatted):

```c++
template <class... _Args>
_LIBCPP_HIDE_FROM_ABI int _LIBCPP_DIAGNOSTIC_PUSH
#ifdef _LIBCPP_COMPILER_CLANG_BASED
_LIBCPP_CLANG_DIAGNOSTIC_IGNORED("-Wgcc-compat") _LIBCPP_ATTRIBUTE_FORMAT(__printf__, 4, 5)
#endif
    __libcpp_snprintf_l(char* __s, size_t __n, locale_t __loc, const char* __format, _Args&&... __args) {
  _LIBCPP_GCC_DIAGNOSTIC_IGNORED("-Wformat-nonliteral")
  return ::snprintf_l(__s, __n, __loc, __format, std::forward<_Args>(__args)...);
}
_LIBCPP_DIAGNOSTIC_POP
```

Comment (philnik777) [reply libcxx/include/__locale_dir/locale_base_api/bsd_locale_defaults.h:89]: Huh. Kinda odd that this is emitted by default. I epxected that we could just introduce `_LIBCPP_VARIADIC_TEMPLATE_FORMAT`. I think it'd be fine if we just wrap all of the functions that use format functions in something like

```c++
_LIBCPP_DIAGNOSTIC_PUSH
_LIBCPP_CLANG_DIAGNOSTIC_IGNORED("-Wgcc-compat")
_LIBCPP_GCC_DIAGNOSTIC_IGNORED("-Wformat-nonliteral") // GCC doesn't support [[gnu::format]] on variadic templates
#ifdef _LIBCPP_COMPILER_CLANG_BASED
#  define _LIBCPP_VARIADTIC_TEMPLATE_FORMAT ...
#else
#  define _LIBCPP_VARIADTIC_TEMPLATE_FORMAT
#endif

<put the functions here>

_LIBCPP_DIAGNOSTIC_POP
```
It's not great, but it may catch a rather nasty bug at some point. We're fallible too after all.

File: libcxx/include/__locale_dir/locale_base_api.h (hunk 2 above)
Comment (philnik777) [reply libcxx/include/__locale_dir/locale_base_api.h:57]: Ah, OK. SGTM. Could you maybe add a comment that this should be refactored?

**Review by philnik777 — APPROVED** [review]

**Comment by ldionne** [comment]
The CI failure is something that failed on `main`, which I fixed separately. Merging!

**Comment by nico** [comment]
This breaks building on mac in some scenarios: https://github.com/llvm/llvm-project/issues/117630
